File: voca_crm_api/src/main/java/com/vocacrm/api/model/Member.java
package com.vocacrm.api.model;

import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.Data;
import lombok.ToString;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.UUID;

/**
 * 회원(Member) 엔티티 클래스
 *
 * VocaCRM 시스템의 회원 정보를 담는 JPA 엔티티입니다.
 * PostgreSQL의 'members' 테이블과 매핑됩니다.
 *
 * 주요 특징:
 * - UUID를 기본키로 사용하여 분산 시스템에서의 확장성 보장
 * - memberNumber는 중복 가능 (동일 회원번호로 여러 레코드 등록 가능)
 * - 생성/수정 시간 자동 관리
 *
 * @author VocaCRM Team
 * @version 1.0
 */
@Data // Lombok: Getter, Setter, toString, equals, hashCode 자동 생성
@Entity // JPA 엔티티임을 표시
@Table(name = "members") // 매핑될 데이터베이스 테이블명 지정
public class Member {

    /**
     * 회원 고유 식별자 (Primary Key)
     */
    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private UUID id;

    /**
     * 회원번호
     * 업무적으로 사용되는 회원번호로, 중복 가능
     * 음성 검색의 주요 검색 키로 사용됨
     */
    @Column(name = "member_number", length = 50, nullable = false)
    private String memberNumber;

    /**
     * 회원 이름
     * 필수 입력 항목 (nullable = false)
     */
    @Column(nullable = false, length = 100)
    private String name;

    /**
     * 전화번호
     * 선택 입력 항목, 하이픈 포함 최대 20자
     * 예: "010-1234-5678"
     */
    @Column(length = 20)
    private String phone;

    /**
     * 이메일 주소
     * 선택 입력 항목, 최대 100자
     */
    @Column(length = 100)
    private String email;

    @Column(name = "business_place_id", length = 7, nullable = false)
    private String businessPlaceId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "business_place_id", insertable = false, updatable = false)
    @ToString.Exclude
    @JsonIgnore
    private BusinessPlace businessPlace;

    /**
     * 회원 소유자 ID (Ownership 기반 권한 체크)
     *
     * 권한 규칙:
     * - 소유자: 수정/삭제 가능
     * - 같은 Role: 수정 가능, 삭제 불가
     * - 상위 Role: 수정/삭제 가능
     */
    @Column(name = "owner_id")
    private UUID ownerId;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "owner_id", insertable = false, updatable = false)
    @ToString.Exclude
    @JsonIgnore
    private User owner;

    /**
     * 마지막으로 수정한 사용자 ID
     */
    @Column(name = "last_modified_by_id")
    private UUID lastModifiedById;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "last_modified_by_id", insertable = false, updatable = false)
    @ToString.Exclude
    @JsonIgnore
    private User lastModifiedBy;

    @Column(length = 20)
    private String grade;

    /**
     * 회원 비고
     * 추가 메모나 특이사항 기록용
     */
    @Column(columnDefinition = "TEXT")
    private String remark;

    /**
     * Soft Delete 여부
     * true인 경우 삭제 대기 상태
     */
    @Column(name = "is_deleted", nullable = false)
    private Boolean isDeleted = false;

    /**
     * 삭제 요청 시간
     */
    @Column(name = "deleted_at")
    private LocalDateTime deletedAt;

    /**
     * 삭제 요청자 ID
     */
    @Column(name = "deleted_by")
    private UUID deletedBy;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "deleted_by", insertable = false, updatable = false)
    @ToString.Exclude
    @JsonIgnore
    private User deletedByUser;

    /**
     * 레코드 생성 시간
     * 자동으로 현재 시간이 설정되며, 이후 변경 불가 (updatable = false)
     */
    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    /**
     * 레코드 최종 수정 시간
     * 엔티티가 업데이트될 때마다 자동으로 현재 시간으로 갱신됨
     */
    @UpdateTimestamp
    @Column(name = "updated_at", nullable = false)
    private LocalDateTime updatedAt;

}

File: voca_crm/lib/domain/entity/member.dart
/// 회원 등급
enum MemberGrade {
  GENERAL,
  VIP;

  String get displayName {
    switch (this) {
      case MemberGrade.GENERAL:
        return '일반';
      case MemberGrade.VIP:
        return 'VIP';
    }
  }

  static MemberGrade? fromString(String? value) {
    if (value == null) return null;
    try {
      return MemberGrade.values.firstWhere(
        (e) => e.name.toLowerCase() == value.toLowerCase(),
      );
    } catch (e) {
      return null;
    }
  }
}

class Member {
  final String id;
  final String memberNumber;
  final String name;
  final String? phone;
  final String? email;
  final String? businessPlaceId;
  final String? ownerId;
  final String? lastModifiedById;
  final String? grade;
  final String? remark;
  final bool isDeleted;
  final DateTime? deletedAt;
  final String? deletedBy;
  final DateTime createdAt;
  final DateTime updatedAt;

  MemberGrade? get gradeEnum => MemberGrade.fromString(grade);

  Member({
    required this.id,
    required this.memberNumber,
    required this.name,
    this.phone,
    this.email,
    this.businessPlaceId,
    this.ownerId,
    this.lastModifiedById,
    this.grade,
    this.remark,
    this.isDeleted = false,
    this.deletedAt,
    this.deletedBy,
    required this.createdAt,
    required this.updatedAt,
  });

  factory Member.fromJson(Map<String, dynamic> json) {
    return Member(
      id: json['id'] as String,
      memberNumber: json['memberNumber'] as String,
      name: json['name'] as String,
      phone: json['phone'] as String?,
      email: json['email'] as String?,
      businessPlaceId: json['businessPlaceId'] as String?,
      ownerId: json['ownerId'] as String?,
      lastModifiedById: json['lastModifiedById'] as String?,
      grade: json['grade'] as String?,
      remark: json['remark'] as String?,
      isDeleted: json['isDeleted'] as bool? ?? false,
      deletedAt: json['deletedAt'] != null
          ? DateTime.parse(json['deletedAt'] as String)
          : null,
      deletedBy: json['deletedBy'] as String?,
      createdAt: json['createdAt'] != null
          ? DateTime.parse(json['createdAt'] as String)
          : DateTime.now(),
      updatedAt: json['updatedAt'] != null
          ? DateTime.parse(json['updatedAt'] as String)
          : DateTime.now(),
    );
  }

  Member copyWith({
    String? id,
    String? memberNumber,
    String? name,
    String? phone,
    String? email,
    String? businessPlaceId,
    String? ownerId,
    String? lastModifiedById,
    String? grade,
    String? remark,
    bool? isDeleted,
    DateTime? deletedAt,
    String? deletedBy,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) {
    return Member(
      id: id ?? this.id,
      memberNumber: memberNumber ?? this.memberNumber,
      name: name ?? this.name,
      phone: phone ?? this.phone,
      email: email ?? this.email,
      businessPlaceId: businessPlaceId ?? this.businessPlaceId,
      ownerId: ownerId ?? this.ownerId,
      lastModifiedById: lastModifiedById ?? this.lastModifiedById,
      grade: grade ?? this.grade,
      remark: remark ?? this.remark,
      isDeleted: isDeleted ?? this.isDeleted,
      deletedAt: deletedAt ?? this.deletedAt,
      deletedBy: deletedBy ?? this.deletedBy,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
    );
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;

    return other is Member &&
        other.id == id &&
        other.memberNumber == memberNumber &&
        other.name == name &&
        other.phone == phone &&
        other.email == email &&
        other.businessPlaceId == businessPlaceId &&
        other.ownerId == ownerId &&
        other.lastModifiedById == lastModifiedById &&
        other.grade == grade &&
        other.remark == remark &&
        other.isDeleted == isDeleted &&
        other.deletedAt == deletedAt &&
        other.deletedBy == deletedBy &&
        other.createdAt == createdAt &&
        other.updatedAt == updatedAt;
  }

  @override
  int get hashCode {
    return id.hashCode ^
        memberNumber.hashCode ^
        name.hashCode ^
        phone.hashCode ^
        email.hashCode ^
        businessPlaceId.hashCode ^
        ownerId.hashCode ^
        lastModifiedById.hashCode ^
        grade.hashCode ^
        remark.hashCode ^
        isDeleted.hashCode ^
        (deletedAt?.hashCode ?? 0) ^
        (deletedBy?.hashCode ?? 0) ^
        createdAt.hashCode ^
        updatedAt.hashCode;
  }
}
